File: typescript/src/ui/overlays/spacebar.ts
import P5 from 'p5';
import Colors from '../colors';
import Fonts from '../fonts';

const WIDTH = 200;
const HEIGHT = 30;

export default class Spacebar {
    private p5: P5;
    private position: P5.Vector;

    constructor(p5: P5, position: P5.Vector) {
        this.p5 = p5;
        this.position = position;
    }

    draw() {
        this.p5.push();

        this.p5.translate(this.position.x, this.position.y);

        this.p5.fill(Colors.BACKGROUND);
        this.p5.stroke(Colors.EDGE);
        this.p5.strokeWeight(1);

        this.p5.translate(-WIDTH / 2, -HEIGHT / 2);

        this.p5.rect(0, 0, WIDTH, HEIGHT, HEIGHT / 5);

        this.p5.fill(Colors.EDGE);
        this.p5.noStroke();
        this.p5.textFont(Fonts.getInstance().fontLight);
        this.p5.textAlign(this.p5.CENTER);
        this.p5.text('SPACEBAR', WIDTH / 2, HEIGHT / 2 + 5);

        this.p5.pop();
    }
}

File: rust/src/overlays/spacebar.rs
use web_sys::CanvasRenderingContext2d;

use crate::utils::{colors::Colors, config::Config, vector::Vector};

pub struct Spacebar {
    position: Vector,
    config: Config,
}

impl Spacebar {
    pub fn new(position: Vector) -> Spacebar {
        let config = Config::new();

        Spacebar { position, config }
    }
}

impl Spacebar {
    pub fn draw(&self, canvas: CanvasRenderingContext2d) {
        canvas.save();

        let width = self.config.overlays.spacebar.width;
        let height = self.config.overlays.spacebar.height;

        let _result = canvas.translate(
            self.position.x - width / 2.0,
            self.position.y - height / 2.0,
        );

        canvas.set_fill_style(&Colors::Background.value().into());
        canvas.fill_rect(0.0, 0.0, width, height);

        canvas.stroke_rect(0.0, 0.0, width, height);

        canvas.set_fill_style(&Colors::Edge.value().into());
        canvas.set_text_align("center");
        canvas.set_font("300 15px 'Exo 2'");
        let _result = canvas.fill_text("SPACEBAR", width / 2.0, height / 2.0 + 5.0);

        canvas.restore();
    }
}
